### ISSUE TITLE
[Epic] Process Instance Modification

### ISSUE THREAD

**Issue by saig0** [camunda/zeebe#9570]
## Description

As a user, I can modify an active process instance to repair the execution. The execution may be stuck and I want to continue the execution on a different activity (i.e. skip or repeat activities).

Use cases:
* Repair a process instance that is stuck (primary)
* Testing activities of a process in isolation (secondary)

Additional resources:
* Related to #9366
* [Kickoff meeting](https://docs.google.com/document/d/1JOHLvFv83ysmiliol0As4-f9nhCeXR6wTlAnG1UvNoQ) (internal)

## Concept

We modify a process instance by **terminating** active element instances and **activating** new element instances.

The termination of an element instance includes:
* writing a `terminating` and `terminated` event 
* deleting event subscriptions
* resolving incidents 
* canceling jobs

The activation of an element instance is similar to #9366 and includes:
* writing a `activate` command
* create variables if defined
* if no element instance of the flow scope exists then activate the flow scope by writing an `activating` and `activated` event
 
### Ancestor selection for activation

This is an advanced feature for activating a new element instance. By default, the new element instance is created within an existing element instance of the flow scope. For example, when activating an element inside an embedded subprocess and the subprocess is already active.

If there is more than one element instance of the flow scope active then the engine can't decide which element instance to create the new element instance in. Instead, the element instance must be selected by its element instance key. The new element instance is created within the selected element instance.

If the selected element instance is not of the flow scope but from a higher scope (e.g. the process instance key instead of the element instance key of the subprocess) then the engine creates a new element instance of the flow scope first and then creates the new element instance within this scope.   

### Propagation of termination

If an element instance is terminated and the element is a container (i.e. an embedded/event subprocess, multi-instance, call activity) then the engine terminates also the child instances (recursively).

The termination is also propagated to higher scopes. After the termination and activation instructions are applied, the engine checks the element instances of container elements. If an element instance of a container element doesn't have any child instances anymore then the engine terminates also the element instance of the container element (recursively).

### Broker

- [x] #9638
- [x] #9639
- [x] #9640
- [x] #9642
- [x] #9643
- [x] #9646 (stretch goal)
- [x] #9663
- [x] #9689
- [x] #9690
- [x] #9691
- [x] #9698
- [x] #9700
- [x] #9701
- [x] #9702
- [ ] #9704
- [x] #9709
- [x] #10078 
- [x] #9976
- [x] #9729
- [x] #9977
- [x] #9980 
- [x] #9982 
- [x] #9983
- [x] #9984 (stretch goal) 
- [x] #9990 (stretch goal)
- [x] #11256
- [x] #9710
- [x] #9728
- [x] #10008 
- [x] #10347

### Gateway API

- [x] #9659
- [x] #9660
- [ ] #9661
- [x] #9662

### Exporters/Metrics

- [x] #9992
- [ ] #9991 (stretch goal)
- [ ] https://github.com/camunda/zeebe-analytics/issues/83
- [x] #9993 (stretch goal)

### Testing

- [ ] Write exhaustive tests for modifying a process instance by activating different elements 
- [ ] Write exhaustive tests for modifying a process instance by terminating different element instances
- [ ] Write randomized process tests for modification (stretch goal)
- [x] #10010
- [ ] #10011 (after release)

### Zeebe Process Test

- [x] https://github.com/camunda/zeebe-process-test/issues/466
- [x] https://github.com/camunda/zeebe-process-test/issues/467 (stretch goal)

### Documentation

- [x] https://github.com/camunda/camunda-platform-docs/issues/1120
- [x] https://github.com/camunda/camunda-platform-docs/issues/1122
- [ ] I can find examples of the modification (stretch goal) 

### Out of scope

- Activate an element inside a multi-instance subprocess
- Activate a start event of the process or an embedded/event subprocess 
- Activate a boundary event
- Activate an element that is attached to an event-based gateway
- Start at a sequence flow 
- Start the element instance after the given element 
- Apply input mappings when creating a flow scope
- Terminate all element instances of an element at once
- Send a modification command using `zbctl`

**Comment by shahamit** [comment]
This is an important task for us to adopt camunda. If there is any way to upvote this issue so that it gets implemented in the next release kindly let me know.

Few questions relevant to our use case -

1. There would be an api in the zeebe client to redo a task (receive/service/user task) that was already completed either by resending the publish command (in case of a receive task) or the complete command (in case of script task) for the given task name/id. Is that a right understanding?
2. Assuming we want to redo a previous task (task B from the attached screenshot), while the process is waiting for Task F. What happens to already completed tasks i.e. from task C to E. Do we get the flexibility skip them and jump to Task F after re-completing Task B?

Thanks.

![image](https://user-images.githubusercontent.com/1281926/175234924-4b3248f3-d28f-4bda-823d-253cd0dfb7bb.png)

**Comment by saig0** [comment]
@shahamit there is maybe a misunderstanding of this feature. The process instance modification is to **repair** a process instance. For example, by canceling the execution of a task, moving it back to a previous task, and continuing from there. Or, by skipping tasks because the underlying services are not available.

Process instance modification should be an exception in the case that something unexpected happens. It is not for implementing a more dynamic process.

If you want more dynamic then you should model this behavior in your process. For example, using a sequence flow that goes back to a previous task, or using some kind of catch event.

---

Regarding your questions:

> There would be an api in the zeebe client to redo a task (receive/service/user task) that was already completed either by resending the publish command (in case of a receive task) or the complete command (in case of script task) for the given task name/id. Is that a right understanding?

Not exactly. There will be no option to redo a specific task. But the API will have the option to terminate active elements (i.e. activated element instances) and to activate new elements. Combining both instructions, it is possible to "jump back in the process". 

> Assuming we want to redo a previous task (task B from the attached screenshot), while the process is waiting for Task F. What happens to already completed tasks i.e. from task C to E. Do we get the flexibility skip them and jump to Task F after re-completing Task B?

No. The process instance continues at the activated elements as usual. In the example, if you activate "Task B" then the process instance waits on "Task B" for a message and continues afterward with "Task C".

**Comment by shahamit** [comment]
> If you want more dynamic then you should model this behavior in your process. For example, using a sequence flow that goes back to a previous task, or using some kind of catch event.

Yes we need this kind of capability but it isn't possible to design a bpmn diagram where every task to have a back link to every other task. In the screenshot above, Task E would back link to Task A,B,C, D and Task F would back link to Task A,B,C,D,E and so on. There was a suggestion to use this capability [when asked on the forum](https://forum.camunda.io/t/updating-user-task-details-after-its-marked-completed/37534/4) for this kind of use case. 

Could process instance modification feature support this use case too?

**Comment by saig0** [comment]
> There was a suggestion to use this capability [when asked on the forum](https://forum.camunda.io/t/updating-user-task-details-after-its-marked-completed/37534/4) for this kind of use case.

I see. But this suggestion was not suitable. It is not recommended to use modification for this use case because the modification feature is powerful but also dangerous. You can break your process instance and it doesn't work well if the process instance executes concurrently.   

> but it isn't possible to design a bpmn diagram where every task to have a back link to every other task

Well, it is possible. And if your use case requires this flow of tasks then you should model it explicitly. 

Maybe, you don't need to support all of these cases. Or, you could model it differently. Feel free to continue the discussion in the forum.

> Could process instance modification feature support this use case too?

No. This is not the way. You should look for a different solution.

**Comment by Zelldon** [comment]
Hey :wave: currently investigating the backlog and closing stale issues, and stumbled over this. I was wondering since we released this feature and this epic is still open. Is any of you still working on this @korthout @saig0 ?
